solr/core/src/java/org/apache/solr/cluster/placement/PlacementPlanFactory.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+import java.util.Set;
+
+/**
+ * Allows plugins to create {@link PlacementPlan}s telling the Solr layer where to create replicas following the processing of
+ * a {@link PlacementRequest}. The Solr layer can (and will) check that the {@link PlacementPlan} conforms to the {@link PlacementRequest} (and
+ * if it does not, the requested operation will fail).
+ */
+public interface PlacementPlanFactory {
+  /**
+   * <p>Creates a {@link PlacementPlan} for adding a new collection and its replicas.
+   *
+   * <p>This is in support of {@link org.apache.solr.cloud.api.collections.CreateCollectionCmd}.
+   */
+  PlacementPlan createPlacementPlanNewCollection(CreateNewCollectionPlacementRequest request, String CollectionName, Set<ReplicaPlacement> replicaPlacements);
+
+  /**
+   * <p>Creates a {@link PlacementPlan} for adding replicas to a given shard of an existing collection.
+   *
+   * <p>This is in support (directly or indirectly) of {@link org.apache.solr.cloud.api.collections.AddReplicaCmd},
+   * {@link org.apache.solr.cloud.api.collections.CreateShardCmd}, {@link org.apache.solr.cloud.api.collections.ReplaceNodeCmd},
+   * {@link org.apache.solr.cloud.api.collections.MoveReplicaCmd}, {@link org.apache.solr.cloud.api.collections.SplitShardCmd},
+   * {@link org.apache.solr.cloud.api.collections.RestoreCmd} and {@link org.apache.solr.cloud.api.collections.MigrateCmd}.
+   * (as well as of {@link org.apache.solr.cloud.api.collections.CreateCollectionCmd} in the specific case of
+   * {@link org.apache.solr.common.params.CollectionAdminParams#WITH_COLLECTION} but this should be removed shortly and
+   * the section in parentheses of this comment should be removed when the {@code withCollection} javadoc link appears broken).
+   */
+  PlacementPlan createPlacementPlanAddReplicas(AddReplicasPlacementRequest request, String CollectionName, Set<ReplicaPlacement> replicaPlacements);
```

Comment:
Should we include MOVEREPLICA here, too? Some placement plans need to know in advance whether it's an unrelated ADD or a part of MOVE operation - eg. if the replica files are on a shared FS then MOVE is not equivalent to ADD + DELETE because the actual files are not being moved.

So I think we need both `createAddReplicas` and `createMoveReplicas` and `createDeleteReplicas`.

Reply:
Is the move replica command picking up the destination or is the destination specified in the API call? If the latter, there will be no call to the placement plugin.
And if the former, the fact that no files are to be moved is relatively transparent to the plugin. The plugin doesn’t do any work but just tells Solr where to put things. Solr code would then either create or move depending on what command it was executing.
The only difference could be move placement computation (if there’s one) should take into account the lower load on the source node (since replicas will be moved off of it).

Reply:
Eh, it looks to me like we need just one method:
`PlacementPlan createPlacementPlan(PlacementRequest request, String collectionName, Set<ReplicaPlacement> replicaPlacements)`

Also ... I'm not sure how useful this interface is. Do we really need to dictate such granular internal API for the plugins? It's the responsibility of `PlacementPlugin` to create a `PlacementPlan`, and it should be free to do it whichever way it wants, not necessarily going through this factory - it seems like an additional hurdle and not a convenience.

Reply:
That was my initial intention but then I switched to the factory approach. We can expose all accessors in PlacementPlan and let the plug-in build instances of it as it wishes. Maybe we should expose all accessors anyway, but I thought rather than letting each plug-in implement these interfaces again it's simpler to provide a factory. Also I saw no real added value from the plug-in perspective to have its own instance type. 
A provided factory allows to introduce constraints on acceptable parameters in a programmer friendly way (what a method accepts, different method names), if Solr side accepts any instance it would have to do some checks at runtime.

Reply:
@sigram I will indeed unify these two requests to a single one. When creating a new collection, the command currently first creates the collection ZK node then does the Assign request, and there's already a DocCollection instance available. Therefore no need to distinguish new collection creation from adding replicas to an existing one. Given the plugin will have to deal with the more complex task of adding replicas (compared to creating a new collection, since positions of existing replicas have to be taken into account), there will be no added work there (minor loss of efficiency for computing placement for new collections, likely will not even be measurable).

Reply:
I meant above I'm unifying the creation requests as well as the placement plans.